sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -416,6 +416,23 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         // these branches can be pruned away
         val (h, t) = branches.span(_._1 != TrueLiteral)
         CaseWhen( h :+ t.head, None)
+
+      case e @ CaseWhen(branches, Some(elseValue)) if {
```

Comment:
can we apply this optimization when there is no elseValue? and have one more case to remove `CaseWhen` if it only has one branch and its value is the same as elseValue.

Reply:
We can not. When no `elseValue`, all the conditions are required to evaluated before hitting the default `elseValue` which is `null`.

Reply:
ah i see. Another optimization is: we can remove branches that have the same the condition. We can do it in next PR.

Reply:
@cloud-fan It sounds like what this #21904 proposes for?